lib/scholar/neighbors/random_projection_forest.ex

```diff
@@ -0,0 +1,363 @@
+        indices = Nx.take_along_axis(indices, level_indices, axis: 1)
+        level_proj = Nx.take_along_axis(level_proj, level_indices, axis: 1)
+
+        right_sizes = Nx.quotient(cell_sizes, 2)
+        left_sizes = right_sizes + Nx.remainder(cell_sizes, 2)
+        cell_sizes = Nx.select(pos < left_sizes, left_sizes, right_sizes)
+        tags = 2 * tags + (pos >= cell_sizes)
+
+        medians =
+          update_medians(
+            pos,
+            left_sizes,
+            right_sizes,
+            level_proj,
+            nodes,
+            width,
+            median_offset,
+            medians
+          )
+
+        pos = Nx.remainder(pos, left_sizes)
+
+        {
+          indices,
+          medians,
+          {tensor, hyperplanes, level + 1, pos, cell_sizes, tags, nodes, 2 * width,
+           2 * median_offset + 1}
+        }
+      end
+
+    {indices, hyperplanes, medians}
+  end
+
+  defnp update_medians(
+          pos,
+          left_sizes,
+          right_sizes,
+          level_proj,
+          nodes,
+          width,
+          median_offset,
+          medians
+        ) do
+    size = Nx.size(pos)
+    {num_trees, num_nodes} = Nx.shape(medians)
+
+    left_mask = pos == left_sizes - 1
+
+    left_indices =
+      Nx.argsort(left_mask, direction: :desc, stable: true, type: :u32)
+      |> Nx.new_axis(0)
+      |> Nx.broadcast({num_trees, size})
+
+    left_first = Nx.take_along_axis(level_proj, left_indices, axis: 1)
+
+    right_mask = pos == right_sizes
+
+    right_indices =
+      Nx.argsort(right_mask, direction: :desc, stable: true, type: :u32)
+      |> Nx.new_axis(0)
+      |> Nx.broadcast({num_trees, size})
+
+    right_first = Nx.take_along_axis(level_proj, right_indices, axis: 1)
+
+    medians_first = (left_first + right_first) / 2
+
+    median_mask = width <= nodes and nodes < width + median_offset
+    median_pos = Nx.argsort(median_mask, direction: :desc, stable: true, type: :u32)
+    level_medians = Nx.take(medians_first, median_pos, axis: 1)
+
+    level_mask =
+      (median_offset <= nodes and nodes < median_offset + width)
+      |> Nx.new_axis(0)
+      |> Nx.broadcast({num_trees, num_nodes})
+
+    Nx.select(
+      level_mask,
+      level_medians,
+      medians
+    )
+  end
+
+  @doc """
+  Computes the leaf indices for every point in the input tensor.
+  If the input tensor contains n points, then the result has shape {n, num_trees, leaf_size}.
+
+  ## Examples
+
+      iex> key = Nx.Random.key(12)
+      iex> tensor = Nx.iota({5, 2})
+      iex> forest = Scholar.Neighbors.RandomProjectionForest.fit(tensor, num_trees: 3, min_leaf_size: 2, key: key)
+      iex> x = Nx.tensor([[3, 4]])
+      iex> Scholar.Neighbors.RandomProjectionForest.predict(forest, x)
+      #Nx.Tensor<
+        u32[1][3][3]
+        [
+          [
+            [0, 1, 2],
+            [0, 1, 2],
+            [4, 3, 2]
+          ]
+        ]
+      >
```

Comment:
How does this relate to a-knn? Or am I missing something and this is not yet the full aKNN implementation?

Reply:
I would separate that from Random Projection forest implementation. The reason is that we should have a separate functions that find nearest neighbors of query points and those that construct k-nn graph of the data.
Like this we can just do `forest = Forest.fit(x, ...)` and `Forest.predict(forest, x)` where `x` has been used both for fitting and prediction. Then we can process the result, which might include removing `i` from `i`-th row (we don't want `i` to be among k-nearest neighbors of itself in a k-NN graph).

Reply:
Actually, I dropped a comment above with my understanding of how this relates to kNN. Should we have a `predict` function that receives the distance function and then traverses the leaves? Or do we leave it as an exercise to the user? If the second, should we provide a code snippet?

Reply:
Oops, we commented at the same time. What you said makes sense. :) I suggested a comment above which I believe is pretty much what you said. :) I will resolve this one.

Reply:
I actually wrote graph.ex that contains functions for creating k-nn graph from data, both exact (using brute-force) and approximate using random projection forest and neighborhood expansion. Thought of submitting a separate PR for that.

Reply:
Another possibility is to really return nearest neighbors in predict function of RandomProjectionForest, but when creating the graph, the user should get `k + 1` neighbors then filter out `i` from `i`-th row.

Reply:
Do you have a link to the paper so I can update our references in #207? :)

Reply:
> Another possibility is to really return nearest neighbors in predict function of RandomProjectionForest

We can totally decide that later and move the code around later.